eclair-core/src/main/scala/fr/acinq/eclair/channel/publish/MempoolTxMonitor.scala

```diff
@@ -0,0 +1,194 @@
+  // @formatter:on
+
+  def apply(nodeParams: NodeParams, bitcoinClient: ExtendedBitcoinClient, loggingInfo: TxPublishInfo): Behavior[Command] = {
+    Behaviors.setup { context =>
+      Behaviors.withMdc(loggingInfo.mdc()) {
+        new MempoolTxMonitor(nodeParams, bitcoinClient, context).start()
+      }
+    }
+  }
+
+}
+
+private class MempoolTxMonitor(nodeParams: NodeParams, bitcoinClient: ExtendedBitcoinClient, context: ActorContext[MempoolTxMonitor.Command])(implicit ec: ExecutionContext = ExecutionContext.Implicits.global) {
+
+  import MempoolTxMonitor._
+
+  private val log = context.log
+
+  def start(): Behavior[Command] = {
+    Behaviors.receiveMessagePartial {
+      case Publish(replyTo, tx, input) => publish(replyTo, tx, input)
+      case Stop => Behaviors.stopped
+    }
+  }
+
+  def publish(replyTo: ActorRef[TxResult], tx: Transaction, input: OutPoint): Behavior[Command] = {
+    context.pipeToSelf(bitcoinClient.publishTransaction(tx)) {
+      case Success(_) => PublishOk
+      case Failure(reason) => PublishFailed(reason)
+    }
+    Behaviors.receiveMessagePartial {
+      case PublishOk =>
+        log.debug("txid={} was successfully published, waiting for confirmation...", tx.txid)
+        waitForConfirmation(replyTo, tx, input)
+      case PublishFailed(reason) if reason.getMessage.contains("rejecting replacement") =>
+        log.info("could not publish tx: a conflicting mempool transaction is already in the mempool")
+        sendResult(replyTo, TxRejected(TxPublisher.ConflictingTxUnconfirmed))
+      case PublishFailed(reason) if reason.getMessage.contains("bad-txns-inputs-missingorspent") =>
+        // This can only happen if one of our inputs is already spent by a confirmed transaction or doesn't exist (e.g.
+        // unconfirmed wallet input that has been replaced).
+        checkInputStatus(input)
+        Behaviors.same
+      case PublishFailed(reason) =>
+        log.error("could not publish transaction", reason)
+        sendResult(replyTo, TxRejected(TxPublisher.UnknownTxFailure))
+      case status: InputStatus =>
+        if (status.spentConfirmed) {
+          log.info("could not publish tx: a conflicting transaction is already confirmed")
+          sendResult(replyTo, TxRejected(TxPublisher.ConflictingTxConfirmed))
+        } else if (status.spentUnconfirmed) {
+          log.info("could not publish tx: a conflicting mempool transaction is already in the mempool")
+          sendResult(replyTo, TxRejected(TxPublisher.ConflictingTxUnconfirmed))
+        } else {
+          log.info("could not publish tx: one of our wallet inputs is not available")
+          sendResult(replyTo, TxRejected(TxPublisher.WalletInputGone))
+        }
+      case CheckInputFailed(reason) =>
+        log.error("could not check input status", reason)
+        sendResult(replyTo, TxRejected(TxPublisher.TxSkipped(retryNextBlock = true))) // we act as if the input is potentially still spendable
```

Comment:
Should the `InputStatus` part be handled in a separate state?

Reply:
We could, but I think it would make the code more complex (because there are two states where we check input status, so I'm not sure how we'd easily specify to what state we should go on success).